### PR TITLE
historyarchive: Update bucket list hash validation for protocol 23

### PR DESCRIPTION
**Pull request by tamirms**
<details>
  <summary>PR Checklist</summary>
  
### PR Structure

* [ ] This PR has reasonably narrow scope (if not, break it down into smaller PRs).
* [ ] This PR avoids mixing refactoring changes with feature changes (split into two PRs
  otherwise).
* [ ] This PR's title starts with name of package that is most changed in the PR, ex.
  `services/friendbot`, or `all` or `doc` if the changes are broad or impact many
  packages.

### Thoroughness

* [ ] This PR adds tests for the most critical parts of the new functionality or fixes.
* [ ] I've updated any docs ([developer docs](https://developers.stellar.org/api/), `.md`
  files, etc... affected by this change). Take a look in the `docs` folder for a given service,
  like [this one](https://github.com/stellar/go/tree/master/services/horizon/internal/docs).

### Release planning

* [ ] I've reviewed the changes in this PR and if I consider them worthwhile for being mentioned on release notes then I have updated the relevant `CHANGELOG.md` within the  component folder structure. For example, if I changed horizon, then I updated ([services/horizon/CHANGELOG.md](services/horizon/CHANGELOG.md). I add a new line item describing the change and reference to this PR. If I don't update a CHANGELOG, I acknowledge this PR's change may not be mentioned in future release notes.  
* [ ] I've decided if this PR requires a new major/minor version according to
  [semver](https://semver.org/), or if it's mainly a patch change. The PR is targeted at the next
  release branch if it's not a patch change.
</details>

### What

Close https://github.com/stellar/go/issues/5425

In protocol 23 there will be two bucketlists: one for the live ledger entries and one for evicted archived ledger entries. Consequently the bucketlist hash in the ledger header will change to (see [CAP-62](https://github.com/stellar/stellar-protocol/blob/master/core/cap-0062.md#changes-to-ledgerheader)):

```
header.bucketListHash = SHA256(liveStateBucketListHash, hotArchiveHash)
```

The checkpoint change reader will need to be updated to calculate the bucket list hash in this way. The [live bucketlist](https://github.com/stellar/stellar-protocol/blob/master/core/cap-0062.md#live-state-bucketlist) will still be the same as the current bucketlist we ingest in the checkpoint change reader so there will not be any changes required there. However, I did add an assertion to ensure the checkpoint change reader is ingesting from the live bucket list instead of the bucket list for evicted ledger entries.

### Known limitations

[N/A]

### CODE REVIEW

File: ingest/checkpoint_change_reader_test.go [hunk 1]
```diff
@@ -399,6 +405,32 @@ func (s *SingleLedgerStateReaderTestSuite) TestMalformedProtocol11BucketNoMeta()
 	s.Assert().Equal("Error while reading from buckets: Read INITENTRY from version <11 bucket: 0@517bea4c6627a688a8ce501febd8c562e737e3d86b29689d9956217640f3c74b", err.Error())
 }
 
+// TestMalformedProtocol11Bucket ensures the checkpoint change reader asserts its reading from the live bucketlist
```
Comment (urvisavla) [inline ingest/checkpoint_change_reader_test.go:408]:
```suggestion
// TestMalformedBucketListType ensures the checkpoint change reader asserts its reading from the live bucketlist
```

File: historyarchive/history_archive_state_test.go [hunk 2]
```diff
@@ -47,112 +48,36 @@ func TestUnmarshalState(t *testing.T) {
...
-            "snap": "0000000000000000000000000000000000000000000000000000000000000000"
-        }
-    ]
-}`)
+	for _, testCase := range []struct {
+		inputFile    string
+		expectedHash string
+	}{
+		{
+			// This is real bucket hash list for pubnet's ledger: 24088895
```
Comment (sreuland) [inline historyarchive/history_archive_state_test.go:56]: nice!

File: ingest/checkpoint_change_reader.go [hunk 3]
```diff
@@ -325,9 +325,18 @@ func (r *CheckpointChangeReader) streamBucketContents(hash historyarchive.Hash,
 				)
 				return false
 			}
-			// We can't use MustMetaEntry() here. Check:
-			// https://github.com/golang/go/issues/32560
-			bucketProtocolVersion = uint32(entry.MetaEntry.LedgerVersion)
+			metaEntry := entry.MustMetaEntry()
+			bucketProtocolVersion = uint32(metaEntry.LedgerVersion)
+			bucketListType, ok := metaEntry.Ext.GetBucketListType()
+			if ok && bucketListType != xdr.BucketListTypeLive {
```
Comment (sreuland) [inline ingest/checkpoint_change_reader.go:331]: what is the real life possibility of this occurrence, can see that only streamBuckets() calls this while parsing currentBuckets, of which that list should only continue to have live buckets correct, if this got triggered it means the remote archive is corrupt, as core would not emit this way, at least that's how it appears in the historyV1.json and historyv2.json test cases and cap-62 didn't mention a behavior change that could lead to this corner case, just trying to understand why this detection is necessary?

Comment (tamirms) [reply ingest/checkpoint_change_reader.go:331]: this code makes sure we are reading the live bucketlist instead of the hot archive bucketlist. in practice this error should never occur unless we are reading off the wrong location in the history archives or if, as you mentioned, the history archives are corrupted.

here is a description of the bucketlists in cap-62 https://github.com/stellar/stellar-protocol/blob/master/core/cap-0062.md#live-state-bucketlist

**Review by sreuland — APPROVED** [review]
